x/twap/api_test.go

```diff
@@ -213,34 +213,181 @@ func (s *TestSuite) TestGetArithmeticTwap() {
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record older than keep threshold); to now; ctxTime = baseTime, start time = end time = before keep threshold": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourBeforeKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourBeforeKeepThreshold, oneHourBeforeKeepThreshold, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record older than keep threshold); with end time; ctxTime = baseTime, start time = end time = before keep threshold - 1ms": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourBeforeKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourBeforeKeepThreshold.Add(-time.Millisecond), oneHourBeforeKeepThreshold.Add(-time.Millisecond), quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record older than keep threshold); to now; ctxTime = after keep threshold, start time = before keep threshold; end time = after keep threshold": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourBeforeKeepThreshold, oneHourAfterKeepThreshold, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record older than keep threshold); with end time; ctxTime = after keep threshold, start time = before keep threshold; end time = after keep threshold - 1ms": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      oneHourAfterKeepThreshold,
+			input:        makeSimpleTwapInput(oneHourBeforeKeepThreshold, oneHourAfterKeepThreshold.Add(-time.Millisecond), quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(1 record at keep threshold); to now; ctxTime = base keep threshold, start time = base time - 1ms (source of error); end time = base keep threshold; error": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTime.Add(-time.Millisecond), baseTimePlusKeepPeriod, quoteAssetA),
+			expectError:  twap.TimeTooOldError{Time: baseTime.Add(-time.Millisecond)},
+		},
+		"(1 record at keep threshold); with end time; ctxTime = base keep threshold, start time = base time - 1ms (source of error); end time = base keep threshold - ms; error": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTime.Add(-time.Millisecond), baseTimePlusKeepPeriod.Add(-time.Millisecond), quoteAssetA),
+			expectError:  twap.TimeTooOldError{Time: baseTime.Add(-time.Millisecond)},
+		},
+		"(2 records); to now; with one directly at threshold, interpolated": {
+			recordsToSet: []types.TwapRecord{baseRecord, recordBeforeKeepThreshold},
+			ctxTime:      baseTimePlusKeepPeriod,
+			input:        makeSimpleTwapInput(baseTime, baseTimePlusKeepPeriod, quoteAssetA),
```

Comment:
Seeing all these test cases, Im thinking if there would be a way to declare all these test cases as variables(including the expected TWAP) in the future

Reply:
These are inputs though and they are constructed based on the logic / setup of the test case. I don't think we should be extracting these since the setup logic varies form test case to test case